From 437e6f9cd705fefeeec4c9f062fd790cecd9584e Mon Sep 17 00:00:00 2001
From: lesssalt <lesssalt@proton.me>
Date: Wed, 4 Sep 2024 21:51:39 +0000
Subject: [PATCH] Add option exit_on_last

When exit_on_last is toggled on, w3m will close tabs and exit if the
user navigates back while on the last buffer in the stack. The use case
is to quit automatically when going back without the need for pressing a
different key or calling an additional command. This setting affects all
calls to backbf.

We change the descriptions of the other quit/close options to be more
accurate.
---
 fm.h   |    1 +
 main.c |    7 ++++---
 rc.c   |    7 +++++--
 3 files changed, 10 insertions(+), 5 deletions(-)

--- a/fm.h
+++ b/fm.h
@@ -971,6 +971,7 @@ global int useVisitedColor init(FALSE);
 global int visited_color init(5);	/* magenta  */
 #endif				/* USE_COLOR */
 global int confirm_on_quit init(TRUE);
+global int exit_on_last init(FALSE);
 #ifdef USE_MARK
 global int use_mark init(FALSE);
 #endif
--- a/main.c
+++ b/main.c
@@ -4234,13 +4234,14 @@ DEFUN(backBf, BACK, "Close current buffe
     Buffer *buf = Currentbuf->linkBuffer[LB_N_FRAME];
 
     if (!checkBackBuffer(Currentbuf)) {
-	if (close_tab_back && nTab >= 1) {
+	if (exit_on_last && nTab == 1)
+	    _quitfm(FALSE);
+	else if (close_tab_back || exit_on_last) {
 	    deleteTab(CurrentTab);
 	    displayBuffer(Currentbuf, B_FORCE_REDRAW);
 	}
 	else
-	    /* FIXME: gettextize? */
-	    disp_message("Can't go back...", TRUE);
+		disp_message(_("Can't go back..."), TRUE);
 	return;
     }
 
--- a/rc.c
+++ b/rc.c
@@ -134,8 +134,9 @@ static int OptionEncode = FALSE;
 #define CMT_PDROOT      N_("Directory corresponding to /~user")
 #define CMT_CGIBIN      N_("Directory corresponding to /cgi-bin")
 #define CMT_TMP         N_("Directory for temporary files")
-#define CMT_CONFIRM_QQ  N_("Confirm when quitting with q")
-#define CMT_CLOSE_TAB_BACK N_("Close tab if buffer is last when back")
+#define CMT_CONFIRM_QQ  N_("Confirm when calling quit")
+#define CMT_EXIT_ON_LAST  N_("Close w3m/tab if buffer is last when going back")
+#define CMT_CLOSE_TAB_BACK N_("Close tab if buffer is last when going back")
 #ifdef USE_MARK
 #define CMT_USE_MARK	N_("Enable mark operations")
 #endif
@@ -518,6 +519,8 @@ struct param_ptr params3[] = {
 #endif				/* USE_HISTORY */
     {"confirm_qq", P_INT, PI_ONOFF, (void *)&confirm_on_quit, CMT_CONFIRM_QQ,
      NULL},
+    {"exit_on_last", P_INT, PI_ONOFF, (void *)&exit_on_last, CMT_EXIT_ON_LAST,
+     NULL},
     {"close_tab_back", P_INT, PI_ONOFF, (void *)&close_tab_back,
      CMT_CLOSE_TAB_BACK, NULL},
 #ifdef USE_MARK
